### PR TITLE
Add nullary function and some unit tests

### LINKED ISSUE #276: extend compute kernel arity to include nullary functions
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

extend compute kernel arity to include nullary functions, so that functions like `random()` can be implemented without too much overhead

**Describe the solution you'd like**

Similar to `unary`, but only takes a usize

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by Jimexist**
# Which issue does this PR close?

We generally require a GitHub issue to be filed for all bug fixes and enhancements and this helps us generate change logs for our releases. You can link an issue to this PR using the GitHub syntax. For example `Closes #123` indicates that this PR will close issue #123.

Closes #276

 # Rationale for this change
 Why are you proposing this change? If this is already explained clearly in the issue then this section is not needed.
 Explaining clearly why changes are proposed helps reviewers understand your changes and offer better suggestions for fixes.  

# What changes are included in this PR?

There is no need to duplicate the description in the issue here but it is sometimes worth providing a summary of the individual changes in this PR.

# Are there any user-facing changes?

If there are user-facing changes then we may require documentation to be updated before approving the PR.

If there are any breaking changes to public APIs, please add the `breaking change` label.

### CODE REVIEW

File: arrow/src/compute/kernels/arity.rs [hunk 1]
```diff
@@ -72,3 +72,40 @@ where
...
 }
+
+/// Applies a nullary and infalible function to generate a primitive array.
+/// You should use this instead of the vectorized version when each generated value is different.
+pub fn nullary<F, O>(size: usize, op: F) -> PrimitiveArray<O>
+where
+    O: ArrowPrimitiveType,
+    F: Fn() -> O::Native,
+{
+    let values = (0..size).map(|_| op());
```
Comment (Dandandan) [inline arrow/src/compute/kernels/arity.rs:83]: This could use `std::iter::repeat`

Comment (Dandandan) [inline arrow/src/compute/kernels/arity.rs:83]: This can use `std::iter::repeat_with`

**Comment by nevi-me** [comment]
@jorgecarleitao @alamb have time to cast second pair of eyes?

**Comment by jorgecarleitao** [comment]
The code itself looks great.

I am not sure this is not shadowing `PrimitiveArray<T>::from_iter_values`, which IMO covers this exact case.

**Comment by Jimexist** [comment]
> The code itself looks great.
> 
> I am not sure this is not shadowing `PrimitiveArray<T>::from_iter_values`, which IMO covers this exact case.

fair point. i've updated the implementation. if this deems unnecessary i'm happy to remove all of the code except for the unit test addition.

File: arrow/src/compute/kernels/arity.rs [hunk 2]
```diff
@@ -72,3 +73,37 @@ where
     let data = into_primitive_array_data::<_, O>(array, buffer);
     PrimitiveArray::<O>::from(data)
 }
+
+/// Applies a nullary and infalible function to generate a primitive array.
+/// You should use this instead of the vectorized version when each generated value is different.
+pub fn nullary<F, O>(size: usize, op: F) -> PrimitiveArray<O>
```
Comment (alamb) [inline arrow/src/compute/kernels/arity.rs:79]: I don't really understand this need for this function either -- `unary` applies a function to an existing array to make a new array.

Given the implementation now, it seems like writing 
```
let arr = Int32Array::from_iter_values(iter::repeat_with(op).take(size)
```

would be just as clear.  

Maybe I am confused about its name. A more specific name might be 'make_constant_array`

I think it takes time to get used to / discover creating `Array`s from iterators -- maybe we can work on adding some more examples to our documentation to help

Comment (Jimexist) [reply arrow/src/compute/kernels/arity.rs:79]: happy to close this PR if not needed.

**Pull request closed without merging**